virttest/cpu.py

```diff
@@ -1258,3 +1261,132 @@ def get_qemu_best_cpu_model(params):
             return host_cpu_model
     # If no host cpu model can be found on qemu_cpu_models, choose the default
     return params.get("default_cpu_model", None)
+
+
+def get_cpu_info_from_cache(cache_file):
+    """
+    Get cpu info from cache_file
+    """
+    if os.path.exists(cache_file):
+        logging.info("Get cpu model and flags from cache file")
+        with open(cache_file) as f:
+            cpu_info = json.load(f)
+        return cpu_info
+    return
+
+
+def store_cpu_info_to_cache(cache_file, cpu_info):
+    """
+    Store cpu info into cache_file
+    """
+    with open(cache_file, "w") as f:
+        json.dump(cpu_info, f)
+    return
+    
+
+def get_qemu_best_cpu_info(params):
+    """
+    Get the best cpu model and features available for qemu
+    """
+    cache_file = "qemu_best_cpu_cache"
+    cache_path = os.path.join(data_dir.get_tmp_dir(), cache_file)
+    cpu_info = get_cpu_info_from_cache(cache_path)
+    if not cpu_info:
+        cpu_info = {}
+        cpu_model = get_qemu_best_cpu_model(params)
+        cpu_features = ""
+        cpu_info["model"] = cpu_model
+        cpu_info["features"] = cpu_features
+        store_cpu_info_to_cache(cache_path, cpu_info)
+    return cpu_info
+
+
+def get_cpu_info_from_virsh(params):
+    """
+    Try to get cpu model and features from virsh with 'host-model'
+    """
+    vm_arch = params["vm_arch_name"]
+    if vm_arch not in ['x86_64', 'i686']:
+        raise NotImplementedError("Arch '%s' is not supported" % vm_arch)
+    machine = params.get('machine_type')
+    name = uuid.uuid4().hex
+    cache_file = "%s_%s_libvirt_host_model_cache" % (vm_arch, machine)
+    cache_path = os.path.join(data_dir.get_tmp_dir(), cache_file)
+    cpu_info = get_cpu_info_from_cache(cache_path)
+    if cpu_info:
+        return cpu_info
+    try:
+        path.find_command("virsh")
+    except path.CmdNotFoundError:
+        logging.warning("Virsh executable not set or found on path")
+        return
+
+    xml = """
+            <domain type='kvm'>
+                <name>%s</name>
+                <memory>1</memory>
+                <os>
+                    <type arch='%s' machine='%s'>hvm</type>
+                </os>
+                <cpu mode='host-model'/>
+            </domain>
+          """ % (name, vm_arch, machine)
+    xml_file = os.path.join(data_dir.get_tmp_dir(), "temp_xml_for_cpu")
```

Comment:
The `xml_file` is not unique if machine changed ?  Suggest use the same strategy for cache_file.

Reply:
It's just a temporary file. We will never reuse it. So I think the file name doesn't matter.

Reply:
Yes, it is true, since use `w` every time. i wonder if this step can be optimized to use an unique xml and reuse it ?
@luckyh What's your opinion about this ?

Reply:
No cache file, no env,  so use fresh xml file each time, ack for this.